### ISSUE TITLE
Port to WebExtensions

### ISSUE THREAD

**Issue by palant** [alexbenenson/transliterator#3]
As of Firefox 57 (release date November 2017), classic extensions will no longer be allowed in Firefox - only [WebExtensions](https://developer.mozilla.org/en-US/Add-ons/WebExtensions) will be supported. The good news: Transliterator can be ported with comparably little effort. Also, the ported version will work on Chrome and Opera as well. The bad news: the extension will no longer apply to the browser's user interface, only to webpages. So while I think that we need to start thinking about the WebExtensions port early, I'd suggest actually releasing that port as late as possible.

What needs to be done:

* The user interface needs to be rewritten from XUL to HTML (this part can be done first, already in the current extension).
* The main functionality needs to be moved into a content script, loading when the hotkey is pressed.
* Some structural and packaging adjustments.

I don't think that migrating the preferences is important for this extension, this simplifies things a lot. However, changing schemes via about:config will no longer be possible - all the more important to finally provide a proper UI for that.

**Comment by alexbenenson** [comment]
Are the chrome(UI) extensions completely going away? One of the more interesting features of transliterator, in my mind, was being able to work on all platform apps - *bird, bluegriffon, etc. And inside multiple popups, such as bookmark manager. And in other extensions. Without that, I don't think it has much to offer over google input tools chrome extension. 

I was hoping to port it to a model where all the logic lives in firefox chrome, and content windows only get messages with replacement characters, similar to how transliterator now sends key strokes to input fields.

**Comment by palant** [comment]
Unfortunately - yes, messing with the Firefox UI isn't going to be possible any more. So if you feel that way you might want to leave the extension in its current state for Thunderbird, SeaMonkey and Bluegriffon only. I doubt that any of them will follow suit and disable classic extensions any time soon.

**Comment by alexbenenson** [comment]
I suppose it's time to start transliterator2.

**Comment by palant** [comment]
Yes, that would be an option as well... Question is, would you care enough to maintain two extensions in parallel? Sadly, I cannot promise to find any time for it beyond preparing the initial WebExtensions release.

**Comment by alexbenenson** [comment]
Fortunately the old version barely needs any maintenance. We'd just have to keep updating targetApplications until there are none left :)

**Comment by palant** [comment]
Then I'd suggest leaving both extensions in the same repository because much of the code will still be identical - regardless of different extension IDs. The legacy extension can live on on the `legacy` branch while `master` will be the WebExtension version.

**Comment by hellboy81** [comment]
Is there alternatives?

**Comment by alexbenenson** [comment]
@palant  - do you know of any way to send trusted KeyEvents from content scripts? This is what's blocking me at the moment.

**Comment by palant** [comment]
Ouch... No, I don't think there is - if the events aren't trusted by default then that's what you are stuck with. I'll file a Firefox bug but given that Chrome's behavior is the same, I wouldn't expect much.

**Comment by palant** [comment]
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1417877

**Comment by alexbenenson** [comment]
Thanks. Let's hope...

**Comment by palant** [comment]
My bug was resolved as a duplicate of https://bugzilla.mozilla.org/show_bug.cgi?id=1348213 which is P5. In other words: very unlikely.

**Comment by alexbenenson** [comment]
bummer. I don't see why an event from a user-installed extension could not be trusted...

**Comment by LmG7119** [comment]
Does this mean we won't get this add-on in FF57.x?

**Comment by palant** [comment]
@alexbenenson Any idea how google input tools works around that issue?

**Comment by LeoG71** [comment]
I believe google input tools is just virtual keyboard installed into windows 10. It does not intercept keystrokes. I could be wrong.

**Comment by palant** [comment]
[This](https://chrome.google.com/webstore/detail/google-input-tools/mclkkofklkfljcocdinagocijmpgbhab) is not a virtual keyboard ;).

I looked into the source code, they seem to be calling `createEvent()`. The only unusual thing is that they create a plain event rather than a keyboard event and set `charCode = 0` and `keyCode = c` on it. Seems to be working for them but probably Chrome-only?

**Comment by alexbenenson** [comment]
I could not find the source code and looked at the minified version... and after all the torture could not reproduce event handling even in chrome. I did see some code manipulating input values / selected ranges directly so maybe that's what they do.

This is an option too, but quite unattractive. Different code paths for form inputs and contentEditable areas, all new kinks and bugs to work out. Not sure I have the energy for this.

**Comment by LmG7119** [comment]
Please Please find that energy. :-)
